basesandbox/tests/intergration_test.rs

```diff
@@ -140,3 +138,91 @@ fn execute_simple_intra_massive() {
         x.join().unwrap();
     }
 }
+
+#[test]
+fn terminator_socket() {
+    let (c1, c2) = DomainSocket::arguments_for_both_ends();
+    let d1 = thread::spawn(|| DomainSocket::new(c1));
+    let d2 = DomainSocket::new(c2);
+    let d1 = d1.join().unwrap();
+    let terminator = d1.create_terminator();
+    let barrier = Arc::new(Barrier::new(2));
+    let barrier_ = barrier.clone();
+    let t = thread::spawn(move || {
+        assert_eq!(d1.recv(None).unwrap(), vec![1, 2, 3]);
+        barrier_.wait();
+        assert_eq!(d1.recv(None).unwrap_err(), cbsb::ipc::RecvError::Termination)
+    });
+    d2.send(&[1, 2, 3]);
+    barrier.wait();
+    terminator.terminate();
+    t.join().unwrap();
+}
+
+#[test]
+fn terminator_intra() {
+    let (c1, c2) = Intra::arguments_for_both_ends();
+    let d1 = thread::spawn(|| Intra::new(c1));
+    let d2 = Intra::new(c2);
+    let d1 = d1.join().unwrap();
+    let terminator = d1.create_terminator();
+    let barrier = Arc::new(Barrier::new(2));
+    let barrier_ = barrier.clone();
+    let t = thread::spawn(move || {
+        assert_eq!(d1.recv(None).unwrap(), vec![1, 2, 3]);
+        barrier_.wait();
+        assert_eq!(d1.recv(None).unwrap_err(), cbsb::ipc::RecvError::Termination)
+    });
+    d2.send(&[1, 2, 3]);
+    barrier.wait();
+    terminator.terminate();
+    t.join().unwrap();
+}
+
+struct TestForward;
+
+impl Forward for TestForward {
+    fn forward(data: &[u8]) -> usize {
+        if data[0] == b'0' {
+            0
+        } else {
+            1
+        }
+    }
+}
+
+#[test]
+fn multiplexer() {
+    let (c1, c2) = DomainSocket::arguments_for_both_ends();
+    let d1 = thread::spawn(|| DomainSocket::new(c1));
+    let d2 = DomainSocket::new(c2);
+    let d1 = d1.join().unwrap();
+
+    let (s1, r1) = d1.split();
+    let (s2, r2) = d2.split();
+
+    let (mut multiplexed1, multiplxer1) = Multiplexer::create::<TestForward, _, _>(s1, r1, 2, 100);
+    let (mut multiplexed2, multiplxer2) = Multiplexer::create::<TestForward, _, _>(s2, r2, 2, 100);
+
+    // multiplxed channel 1
+    let (s1_1, r1_1) = multiplexed1.pop().unwrap();
+    let (s2_1, r2_1) = multiplexed2.pop().unwrap();
+
+    // multiplxed channel 0
+    let (s1_2, r1_2) = multiplexed1.pop().unwrap();
+    let (s2_2, r2_2) = multiplexed2.pop().unwrap();
+
+    s1_1.send(b"11".to_vec()).unwrap();
+    s2_1.send(b"12".to_vec()).unwrap();
+    s2_2.send(b"03".to_vec()).unwrap();
+    s1_2.send(b"04".to_vec()).unwrap();
+
+    assert_eq!(r2_1.recv().unwrap(), b"11");
+    assert_eq!(r1_2.recv().unwrap(), b"03");
+    assert_eq!(r2_2.recv().unwrap(), b"04");
+    assert_eq!(r1_1.recv().unwrap(), b"12");
+
+    // we have to drop multiplexer itself first
+    drop(multiplxer1);
+    drop(multiplxer2);
```

Comment:
How about changing indexes a bit?

```suggestion
    let (mut multiplexed1, multiplxer1) = Multiplexer::create::<TestForward, _, _>(s1, r1, 2, 100);
    let (mut multiplexed2, multiplxer2) = Multiplexer::create::<TestForward, _, _>(s2, r2, 2, 100);

    // multiplxed channel 1
    let (s1_1, r1_1) = multiplexed1.pop().unwrap();
    let (s2_1, r2_1) = multiplexed2.pop().unwrap();

    // multiplxed channel 0
    let (s1_0, r1_0) = multiplexed1.pop().unwrap();
    let (s2_0, r2_0) = multiplexed2.pop().unwrap();

    s1_1.send(b"11".to_vec()).unwrap();
    assert_eq!(r2_1.recv().unwrap(), b"11");

    s2_1.send(b"12".to_vec()).unwrap();
    assert_eq!(r1_1.recv().unwrap(), b"12");
    
    s2_0.send(b"03".to_vec()).unwrap();
    assert_eq!(r1_0.recv().unwrap(), b"03");
    
    s1_0.send(b"04".to_vec()).unwrap();
    assert_eq!(r2_0.recv().unwrap(), b"04");
    
    // we have to drop multiplexer itself first
    drop(multiplxer1);
    drop(multiplxer2);
```